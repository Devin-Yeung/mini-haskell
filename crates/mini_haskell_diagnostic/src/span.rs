use ariadne::Label;

#[derive(Debug, PartialEq, Clone)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl From<logos::Span> for Span {
    fn from(value: logos::Span) -> Self {
        Span {
            start: value.start,
            end: value.end,
        }
    }
}

impl From<Span> for ariadne::Label {
    fn from(val: Span) -> Self {
        Label::new(val.start..val.end)
    }
}
