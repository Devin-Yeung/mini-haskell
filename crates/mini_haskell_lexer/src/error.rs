use mini_haskell_diagnostic::diagnostic::Diagnostic;
use mini_haskell_diagnostic::span::Span;
use thiserror::Error;

#[derive(Error, Debug, PartialEq, Clone)]
#[non_exhaustive]
pub enum LexingError {
    #[error("Unexpected token")]
    UnexpectedToken(Span),
}

impl Diagnostic for LexingError {
    fn span(&self) -> Span {
        match self {
            LexingError::UnexpectedToken(span) => span.clone(),
        }
    }

    fn message(&self) -> String {
        self.to_string()
    }
}
