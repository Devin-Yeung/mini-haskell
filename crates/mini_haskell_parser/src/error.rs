use miette::Diagnostic;
use mini_haskell_diagnostic::span::Span;
use thiserror::Error;

#[derive(Diagnostic, Error, Debug, Eq, PartialEq, Copy, Clone)]
pub enum SyntaxError {
    #[error("Unexpected EOF")]
    UnexpectedEOF,
    #[error("SyntaxError: Unexpected token")]
    UnexpectedToken {
        #[label = "Expected {expected}, found {found}"]
        span: Span,
        expected: &'static str,
        found: &'static str,
    },
    #[error("Expected {0}")]
    Expected(&'static str),
}
