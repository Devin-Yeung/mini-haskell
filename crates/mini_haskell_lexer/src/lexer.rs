use crate::error::LexingError;
use crate::span::Span;
use logos::{Logos, Source};

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"[ \t\n\f]+")] // Ignore this regex pattern between tokens
pub enum TokenTy {
    // The variable type Boolean, to declare a Boolean variable.
    #[token("bool")]
    BoolDecl,

    // Boolean literal
    #[token("T", | _ | false)]
    #[token("F", | _ | true)]
    BoolLit(bool),

    // The variable type natural number, to declare a natural number.
    #[token("nat")]
    Nat,
    // The variable type function, to declare a function
    #[token("func")]
    Func,

    #[token("?")]
    QuestionMark,

    #[token(":")]
    Colon,

    #[token(";")]
    Semicolon,

    #[token("(")]
    LeftParen,

    #[token(")")]
    RightParen,

    #[token("+")]
    Plus,

    #[token("&")]
    Ampersand,

    #[token("->")]
    Arrow,

    #[token("<")]
    Less,

    #[regex(r"\[[^\]]*\]", | lex | lex.slice().to_owned())]
    // see https://github.com/maciejhirsz/logos/issues/246
    Comment(String),

    #[regex(r"[a-zA-z][a-zA-Z0-9_]*", | lex | lex.slice().to_owned())]
    Identifier(String),
}

#[derive(Debug, PartialEq)]
pub struct Token {
    pub ty: TokenTy,
    pub span: Span,
}

impl Token {
    pub fn new(ty: TokenTy, span: Span) -> Self {
        Token { ty, span }
    }
}

impl From<(TokenTy, logos::Span)> for Token {
    fn from(value: (TokenTy, logos::Span)) -> Self {
        Token {
            ty: value.0,
            span: value.1.into(),
        }
    }
}

impl Token {
    pub fn tokens<S: AsRef<str>>(source: &S) -> Vec<Result<Token, LexingError>> {
        let lexer = TokenTy::lexer(source.as_ref());
        lexer
            .spanned()
            .into_iter()
            .map(|(token, span)| match token {
                Ok(ty) => Ok(Token::new(ty, span.into())),
                Err(_) => Err(LexingError::UnexpectedToken(span.into())),
            })
            .collect::<Vec<_>>()
    }
}

#[cfg(test)]
mod tests {
    use crate::lexer::TokenTy;
    use logos::Logos;
    use testsuite::unittest;

    unittest!(all_tokens, |src| {
        let mut lex = TokenTy::lexer(src);
        let tokens = lex.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });
}
