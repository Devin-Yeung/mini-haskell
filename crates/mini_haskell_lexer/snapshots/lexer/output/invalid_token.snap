---
source: crates/mini_haskell_lexer/src/lexer.rs
expression: tokens
---
[
    Err(
        UnexpectedToken(
            Span {
                start: 0,
                end: 1,
            },
        ),
    ),
]
